File: backend/internal/handler/order.go
package handler

import (
	"fmt"
	"math"
	"net/http"
	"time"

	"github.com/google/uuid"
	"github.com/vibestore/backend/internal/dto"
	"github.com/vibestore/backend/internal/middleware"
	"github.com/vibestore/backend/internal/model"
	"github.com/vibestore/backend/internal/repository"
)

type OrderHandler struct {
	orderRepo   *repository.OrderRepository
	productRepo *repository.ProductRepository
}

func NewOrderHandler(orderRepo *repository.OrderRepository, productRepo *repository.ProductRepository) *OrderHandler {
	return &OrderHandler{orderRepo: orderRepo, productRepo: productRepo}
}

// POST /api/v1/store/cart - Validate cart items
func (h *OrderHandler) ValidateCart(w http.ResponseWriter, r *http.Request) {
	storeID := middleware.GetStoreID(r.Context())

	var req dto.CartValidationRequest
	if err := decodeJSON(r, &req); err != nil {
		writeError(w, http.StatusBadRequest, "VALIDATION_ERROR", "Invalid request body")
		return
	}

	allValid := true
	var items []dto.CartValidatedItem
	for _, item := range req.Items {
		product, err := h.productRepo.FindByID(r.Context(), item.ProductID, storeID)
		if err != nil || product == nil || !product.IsActive {
			items = append(items, dto.CartValidatedItem{
				ProductID: item.ProductID,
				Quantity:  item.Quantity,
				Available: false,
			})
			allValid = false
			continue
		}

		available := product.Stock >= item.Quantity
		if !available {
			allValid = false
		}

		var imgURL *string
		if len(product.Images) > 0 {
			imgURL = &product.Images[0].URL
		}

		items = append(items, dto.CartValidatedItem{
			ProductID:    product.ID,
			ProductName:  product.Name,
			ProductImage: imgURL,
			Price:        product.Price,
			Quantity:     item.Quantity,
			Stock:        product.Stock,
			Available:    available,
		})
	}

	writeJSON(w, http.StatusOK, dto.CartValidationResponse{
		Items: items,
		Valid: allValid,
	})
}

// POST /api/v1/store/orders - Checkout
func (h *OrderHandler) Checkout(w http.ResponseWriter, r *http.Request) {
	storeID := middleware.GetStoreID(r.Context())

	var req dto.CheckoutRequest
	if err := decodeJSON(r, &req); err != nil {
		writeError(w, http.StatusBadRequest, "VALIDATION_ERROR", "Invalid request body")
		return
	}

	if req.CustomerName == "" || req.CustomerEmail == "" || len(req.Items) == 0 {
		writeError(w, http.StatusBadRequest, "VALIDATION_ERROR", "Name, email, and items are required")
		return
	}

	// Validate delivery method
	validDeliveryMethods := map[string]bool{"pickup": true, "delivery": true}
	if req.DeliveryMethod == "" {
		req.DeliveryMethod = "delivery"
	}
	if !validDeliveryMethods[req.DeliveryMethod] {
		writeError(w, http.StatusBadRequest, "VALIDATION_ERROR", "Invalid delivery method")
		return
	}

	// Validate payment method
	validPaymentMethods := map[string]bool{"credit_card": true, "debit_card": true, "cash": true, "pix": true, "pay_on_pickup": true}
	if req.DeliveryMethod == "pickup" {
		req.PaymentMethod = "pay_on_pickup"
	}
	if !validPaymentMethods[req.PaymentMethod] {
		writeError(w, http.StatusBadRequest, "VALIDATION_ERROR", "Invalid payment method")
		return
	}

	// If delivery, require shipping address fields
	if req.DeliveryMethod == "delivery" {
		if req.ShippingAddress == "" || req.ShippingCity == "" || req.ShippingState == "" || req.ShippingZip == "" {
			writeError(w, http.StatusBadRequest, "VALIDATION_ERROR", "Shipping address is required for delivery")
			return
		}
	}

	// Build order items and calculate totals
	var orderItems []model.OrderItem
	var subtotal float64

	for _, item := range req.Items {
		product, err := h.productRepo.FindByID(r.Context(), item.ProductID, storeID)
		if err != nil || product == nil || !product.IsActive {
			writeError(w, http.StatusBadRequest, "VALIDATION_ERROR", fmt.Sprintf("Product %d not found or inactive", item.ProductID))
			return
		}
		if product.Stock < item.Quantity {
			writeError(w, http.StatusBadRequest, "VALIDATION_ERROR", fmt.Sprintf("Insufficient stock for %s", product.Name))
			return
		}

		itemTotal := product.Price * float64(item.Quantity)
		subtotal += itemTotal

		var imgURL *string
		if len(product.Images) > 0 {
			imgURL = &product.Images[0].URL
		}

		pid := product.ID
		orderItems = append(orderItems, model.OrderItem{
			ProductID:    &pid,
			ProductName:  product.Name,
			ProductImage: imgURL,
			Price:        product.Price,
			Quantity:     item.Quantity,
			Total:        itemTotal,
		})
	}

	orderNumber := fmt.Sprintf("VS-%s-%s", time.Now().Format("20060102"), uuid.New().String()[:8])

	// Validate change_for if cash payment
	var changeFor *float64
	if req.PaymentMethod == "cash" && req.ChangeFor != nil {
		if *req.ChangeFor < subtotal {
			writeError(w, http.StatusBadRequest, "VALIDATION_ERROR", "Change amount must be greater than or equal to order total")
			return
		}
		changeFor = req.ChangeFor
	}

	order := &model.Order{
		StoreID:         storeID,
		OrderNumber:     orderNumber,
		Status:          "pending",
		CustomerName:    req.CustomerName,
		CustomerEmail:   req.CustomerEmail,
		CustomerPhone:   req.CustomerPhone,
		ShippingAddress: req.ShippingAddress,
		ShippingCity:    req.ShippingCity,
		ShippingState:   req.ShippingState,
		ShippingZip:     req.ShippingZip,
		Subtotal:        subtotal,
		ShippingCost:    0,
		Discount:        0,
		Total:           subtotal,
		Notes:           req.Notes,
		DeliveryMethod:  req.DeliveryMethod,
		PaymentMethod:   req.PaymentMethod,
		ChangeFor:       changeFor,
		Items:           orderItems,
	}

	if err := h.orderRepo.Create(r.Context(), order); err != nil {
		writeError(w, http.StatusInternalServerError, "INTERNAL_ERROR", "Failed to create order")
		return
	}

	writeJSON(w, http.StatusCreated, order)
}

// GET /api/v1/store/orders/{orderNumber}
func (h *OrderHandler) GetByOrderNumber(w http.ResponseWriter, r *http.Request) {
	storeID := middleware.GetStoreID(r.Context())
	orderNumber := r.URL.Query().Get("order_number")
	if orderNumber == "" {
		writeError(w, http.StatusBadRequest, "VALIDATION_ERROR", "Order number is required")
		return
	}

	order, err := h.orderRepo.FindByOrderNumber(r.Context(), orderNumber, storeID)
	if err != nil || order == nil {
		writeError(w, http.StatusNotFound, "NOT_FOUND", "Order not found")
		return
	}

	writeJSON(w, http.StatusOK, order)
}

// GET /api/v1/admin/orders
func (h *OrderHandler) AdminList(w http.ResponseWriter, r *http.Request) {
	storeID := middleware.GetAdminStoreID(r.Context())
	page := queryInt(r, "page", 1)
	perPage := queryInt(r, "per_page", 20)
	status := r.URL.Query().Get("status")
	search := r.URL.Query().Get("search")

	orders, total, err := h.orderRepo.FindByStoreID(r.Context(), storeID, page, perPage, status, search)
	if err != nil {
		writeError(w, http.StatusInternalServerError, "INTERNAL_ERROR", "Failed to list orders")
		return
	}
	if orders == nil {
		orders = []model.Order{}
	}

	writeJSON(w, http.StatusOK, dto.PaginatedResponse{
		Data:       orders,
		Total:      total,
		Page:       page,
		PerPage:    perPage,
		TotalPages: int(math.Ceil(float64(total) / float64(perPage))),
	})
}

// GET /api/v1/admin/orders/{id}
func (h *OrderHandler) AdminGet(w http.ResponseWriter, r *http.Request) {
	storeID := middleware.GetAdminStoreID(r.Context())
	id, err := urlParamUint64(r, "id")
	if err != nil {
		writeError(w, http.StatusBadRequest, "VALIDATION_ERROR", "Invalid order ID")
		return
	}

	order, err := h.orderRepo.FindByID(r.Context(), id, storeID)
	if err != nil || order == nil {
		writeError(w, http.StatusNotFound, "NOT_FOUND", "Order not found")
		return
	}

	writeJSON(w, http.StatusOK, order)
}

// PUT /api/v1/admin/orders/{id}/status
func (h *OrderHandler) AdminUpdateStatus(w http.ResponseWriter, r *http.Request) {
	storeID := middleware.GetAdminStoreID(r.Context())
	id, err := urlParamUint64(r, "id")
	if err != nil {
		writeError(w, http.StatusBadRequest, "VALIDATION_ERROR", "Invalid order ID")
		return
	}

	var req dto.UpdateOrderStatusRequest
	if err := decodeJSON(r, &req); err != nil {
		writeError(w, http.StatusBadRequest, "VALIDATION_ERROR", "Invalid request body")
		return
	}

	validStatuses := map[string]bool{
		"pending": true, "confirmed": true, "processing": true,
		"shipped": true, "delivered": true, "cancelled": true,
	}
	if !validStatuses[req.Status] {
		writeError(w, http.StatusBadRequest, "VALIDATION_ERROR", "Invalid status")
		return
	}

	if err := h.orderRepo.UpdateStatus(r.Context(), id, storeID, req.Status); err != nil {
		writeError(w, http.StatusInternalServerError, "INTERNAL_ERROR", "Failed to update order status")
		return
	}

	order, _ := h.orderRepo.FindByID(r.Context(), id, storeID)
	writeJSON(w, http.StatusOK, order)
}

File: backend-bun/src/routes/order.ts
import Elysia from "elysia";
import { tenantMiddleware } from "../middleware/tenant";
import { authMiddleware } from "../middleware/auth";
import { rateLimit } from "../middleware/rateLimit";
import * as orderRepo from "../repositories/order";
import * as productRepo from "../repositories/product";
import * as storeRepo from "../repositories/store";
import * as adminRepo from "../repositories/admin";
import { sendOrderConfirmation, sendStatusUpdate, notifyStoreNewOrder } from "../services/email";
import { logger } from "../logger";
import type { CartValidationRequest, CheckoutRequest, UpdateOrderStatusRequest, CartValidatedItem } from "../types";

export const publicOrderRoutes = new Elysia({ prefix: "/api/v1" })
  .use(tenantMiddleware)
  .use(rateLimit({ name: "checkout", max: 5, windowSec: 60 }))
  .post("/store/cart", async ({ body, storeId }) => {
    const req = body as CartValidationRequest;
    const items: CartValidatedItem[] = [];
    let valid = true;

    for (const cartItem of req.items) {
      const product = await productRepo.findProductByID(cartItem.product_id, storeId);
      if (!product) {
        items.push({
          product_id: cartItem.product_id,
          product_name: "Unknown",
          product_image: null,
          price: 0,
          quantity: cartItem.quantity,
          stock: 0,
          available: false,
        });
        valid = false;
        continue;
      }

      const available = product.is_active && product.stock >= cartItem.quantity;
      if (!available) valid = false;

      items.push({
        product_id: product.id,
        product_name: product.name,
        product_image: product.images && product.images.length > 0 ? product.images[0].url : null,
        price: product.price,
        quantity: cartItem.quantity,
        stock: product.stock,
        available,
      });
    }

    return { items, valid };
  })
  .post("/store/orders", async ({ body, storeId, set }) => {
    const req = body as CheckoutRequest;

    // Validate required fields
    if (!req.customer_name || !req.customer_email || !req.items || req.items.length === 0) {
      set.status = 400;
      return { code: "VALIDATION_ERROR", message: "customer_name, customer_email, and items are required" };
    }

    // LGPD: require explicit privacy consent
    if (!req.privacy_accepted) {
      set.status = 400;
      return { code: "VALIDATION_ERROR", message: "Aceite da Política de Privacidade é obrigatório" };
    }

    // Validate delivery/payment methods
    const validDelivery = ["pickup", "delivery"];
    const validPayment = ["credit_card", "debit_card", "cash", "pix", "pay_on_pickup"];
    let deliveryMethod = req.delivery_method || "delivery";
    let paymentMethod = req.payment_method || "pix";

    if (!validDelivery.includes(deliveryMethod)) {
      set.status = 400;
      return { code: "VALIDATION_ERROR", message: "Invalid delivery method" };
    }
    if (!validPayment.includes(paymentMethod)) {
      set.status = 400;
      return { code: "VALIDATION_ERROR", message: "Invalid payment method" };
    }

    // Pickup forces pay_on_pickup
    if (deliveryMethod === "pickup") {
      paymentMethod = "pay_on_pickup";
    }

    // Delivery requires shipping fields
    if (deliveryMethod === "delivery") {
      if (!req.shipping_address || !req.shipping_city || !req.shipping_state || !req.shipping_zip) {
        set.status = 400;
        return { code: "VALIDATION_ERROR", message: "Shipping address is required for delivery" };
      }
    }

    // Validate stock and build order items
    let subtotal = 0;
    const orderItems: { product_id: number; product_name: string; product_image: string | null; price: number; quantity: number }[] = [];

    for (const cartItem of req.items) {
      const product = await productRepo.findProductByID(cartItem.product_id, storeId);
      if (!product) {
        set.status = 400;
        return { code: "VALIDATION_ERROR", message: `Product ${cartItem.product_id} not found` };
      }
      if (product.stock < cartItem.quantity) {
        set.status = 400;
        return { code: "VALIDATION_ERROR", message: `Insufficient stock for ${product.name}` };
      }

      const imgUrl = product.images && product.images.length > 0 ? product.images[0].url : null;
      orderItems.push({
        product_id: product.id,
        product_name: product.name,
        product_image: imgUrl,
        price: product.price,
        quantity: cartItem.quantity,
      });
      subtotal += product.price * cartItem.quantity;
    }

    subtotal = Math.round(subtotal * 100) / 100;
    const total = subtotal; // No shipping cost or discount from client

    // Validate change_for
    if (paymentMethod === "cash" && req.change_for !== undefined && req.change_for !== null) {
      if (req.change_for < subtotal) {
        set.status = 400;
        return { code: "VALIDATION_ERROR", message: "Change amount must be >= subtotal" };
      }
    }

    // Generate order number
    const now = new Date();
    const dateStr = now.toISOString().slice(0, 10).replace(/-/g, "");
    const uuid8 = crypto.randomUUID().slice(0, 8);
    const orderNumber = `PS-${dateStr}-${uuid8}`;

    try {
      const orderId = await orderRepo.createOrder(
        {
          store_id: storeId,
          order_number: orderNumber,
          status: "pending",
          customer_name: req.customer_name,
          customer_email: req.customer_email,
          customer_phone: req.customer_phone ?? null,
          shipping_address: req.shipping_address || "",
          shipping_city: req.shipping_city || "",
          shipping_state: req.shipping_state || "",
          shipping_zip: req.shipping_zip || "",
          subtotal,
          shipping_cost: 0,
          discount: 0,
          total,
          notes: req.notes ?? null,
          delivery_method: deliveryMethod,
          payment_method: paymentMethod,
          change_for: req.change_for ?? null,
        },
        orderItems
      );

      const order = await orderRepo.findOrderByID(orderId, storeId);
      set.status = 201;

      // Send emails (fire-and-forget, parallel)
      if (order) {
        Promise.all([
          storeRepo.findStoreByID(storeId),
          adminRepo.findAdminEmailsByStoreID(storeId),
        ]).then(([store, adminEmails]) => {
          const storeName = store?.name ?? "ProtiStore";
          sendOrderConfirmation(order, storeName).catch(() => {});
          notifyStoreNewOrder(order, storeName, adminEmails).catch(() => {});
        }).catch(() => {});
      }

      return order;
    } catch (err: unknown) {
      logger.error(err, "Failed to create order");
      set.status = 500;
      return { code: "INTERNAL_ERROR", message: "Erro ao criar pedido. Tente novamente." };
    }
  })
  .get("/store/orders", async ({ query, storeId, set }) => {
    const orderNumber = query.order_number as string;
    if (!orderNumber) {
      set.status = 400;
      return { code: "VALIDATION_ERROR", message: "order_number query parameter is required" };
    }
    const order = await orderRepo.findOrderByOrderNumber(orderNumber, storeId);
    if (!order) {
      set.status = 404;
      return { code: "NOT_FOUND", message: "Order not found" };
    }
    // Public tracking: return only non-sensitive fields
    return {
      order_number: order.order_number,
      status: order.status,
      delivery_method: order.delivery_method,
      created_at: order.created_at,
      updated_at: order.updated_at,
    };
  });

export const adminOrderRoutes = new Elysia({ prefix: "/api/v1/admin" })
  .use(tenantMiddleware)
  .use(authMiddleware)
  .get("/orders", async ({ query, storeId }) => {
    const page = Number(query.page) || 1;
    const perPage = Number(query.per_page) || 20;
    const status = (query.status as string) || undefined;
    const search = (query.search as string) || undefined;

    const { orders, total } = await orderRepo.findOrdersByStoreID(storeId, page, perPage, status, search);
    return {
      data: orders,
      total,
      page,
      per_page: perPage,
      total_pages: Math.ceil(total / perPage),
    };
  })
  .get("/orders/:id", async ({ params, storeId, set }) => {
    const order = await orderRepo.findOrderByID(Number(params.id), storeId);
    if (!order) {
      set.status = 404;
      return { code: "NOT_FOUND", message: "Order not found" };
    }
    return order;
  })
  .put("/orders/:id/status", async ({ params, body, storeId, set }) => {
    const req = body as UpdateOrderStatusRequest;
    const validStatuses = ["pending", "confirmed", "processing", "shipped", "delivered", "cancelled"];
    if (!validStatuses.includes(req.status)) {
      set.status = 400;
      return { code: "VALIDATION_ERROR", message: "Invalid status" };
    }

    const order = await orderRepo.findOrderByID(Number(params.id), storeId);
    if (!order) {
      set.status = 404;
      return { code: "NOT_FOUND", message: "Order not found" };
    }

    await orderRepo.updateOrderStatus(order.id, storeId, req.status);
    const updated = await orderRepo.findOrderByID(order.id, storeId);

    // Send status update email (fire-and-forget)
    if (updated) {
      const store = await storeRepo.findStoreByID(storeId);
      sendStatusUpdate(updated, store?.name ?? "ProtiStore").catch(() => {});
    }

    return updated;
  });
